### PR TITLE
Move towards making most public classes Final

### PR DESCRIPTION
**Pull request by njsmith**
This changes most of Trio's public classes so that attempts to
subclass them will produce a deprecation warning, with the idea that
once the deprecation period passes on we'll use metaclass=Final to
make subclassing an error.

See #1104 for discussion.

There are some special cases where we do allow subclassing: ABCs,
exceptions, enums, etc. I also added a test to make sure that every
public class is either Final, on its way to becoming Final, or else
falls into one of the special cases. The idea is that when we add new
classes in the future, this will force us to either mark them Final,
or else add a new special case, hopefully with a comment explaining
what makes the new class a special case.

### CODE REVIEW

File: newsfragments/1044.removal.rst [hunk 1]
```diff
@@ -0,0 +1,17 @@
+Most of the public classes that Trio exports – like `trio.Lock`,
+`trio.SocketStream`, and so on – weren't designed with subclassing in
+mind. And we've noticed that some users were trying to subclass them
+anyway, and ending up with fragile code that we're likely to
+accidentally break in the future, or else be stuck unable to make
+changes for fear of breaking subclasses.
+
+There are also some classes that were explicitly designed to be
+subclassed, like the ones in `trio.abc`. Subclassing these is still
```
Comment (pquentin) [inline newsfragments/1044.removal.rst:9]: For what it's worth, the docs check is failing with `/home/runner/work/trio/trio/docs/source/history.rst:37:py:obj reference target not found: trio.abc`

Comment (njsmith) [reply newsfragments/1044.removal.rst:9]: Ah, whoops, thanks for the heads up! I forgot we don't have a dedicated page for `trio.abc`.

File: trio/_highlevel_generic.py [hunk 2]
```diff
@@ -35,7 +37,9 @@ async def aclose_forcefully(resource):
 
 
 @attr.s(eq=False, hash=False)
-class StapledStream(HalfCloseableStream):
+class StapledStream(
```
Comment (oremanj) [inline trio/_highlevel_generic.py:40]: I think StapledStream is safe enough to allow subclassing of, since it has no private attributes.

Comment (njsmith) [reply trio/_highlevel_generic.py:40]: Yeah, the cost of supporting subclasses would be relatively low compared to other things. But that's only the "cost" side of the cost/benefit analysis :-). Is there some reason `StapledStream` subclassing is particularly valuable? I'd rather keep things simple until there's a good reason not too.

Comment (oremanj) [reply trio/_highlevel_generic.py:40]: I don't have an immediate use case and am happy to leave it non-subclassable for consistency.

Comment (njsmith) [reply trio/_highlevel_generic.py:40]: :+1: We can always change it later.

**Pull request merged**